wurst/network/Network.wurst

```diff
@@ -0,0 +1,506 @@
+
+		GamecacheKeys provides int-string conversion for keys for usage in gamecaches.
+
+		GamecacheBuffer is a gamecache-based container for writing
+		integers, reals and booleans. There is a GamecacheBuffer for each primitive type,
+		int, bool, real and asciiInts.
+
+		Network is the main class that coordinates HashBuffer and GamecacheBuffer and does
+		all the heavy lifting.
+
+		Before starting the transmission, the HashBuffer is locked into an immutable state, in
+		which incorrect mutation attempts will print warnings, so long as safety checks are not disabled.
+		The maximum amount of data across all primitive buffer is calculated, and the amount of 
+		required 'sync rounds' is calculated - that is, the amount of times we need to flush/sync
+		data out of the gamecaches to keep key sizes short.
+
+		Since only the local player has any knowledge about the amount of data needed to be sent,
+		and consequently, the amount of sync rounds required, we first send a pre-fetch "metadata"
+		payload with the amount of data in each buffer and the amount of sync rounds, using fixed
+		keys. At the same time, we also send the first payload.
+
+		When a round is received, we write data to the HashBuffer, using the metadata to know when
+		to stop, and start another sync round if necessary. If it is not necessary,
+		we open the HashBuffer for reading and call the finish callback, and destroy the instance.
+********************************************************************************/
+
+enum NetworkState
+	PREPARING
+	SENDING_ROUND
+	FINISHED
+
+interface NetworkFinishedCallback
+	function onFinish(HashBuffer buffer)
+
+public class Network
+	// gamecache used for syncing integers, reals and booleans
+	private static constant DATA_CACHE   = InitGameCache("i.w3v")
+	// gamecache used for syncing strings (as int sequences)
+	private static constant STRING_CACHE = InitGameCache("s.w3v")
+
+	private var currentState = NetworkState.PREPARING
+
+	// mission key to be used for this instance in all game caches
+	private var mkey = GamecacheKeys.getKey(this castTo int)
+
+	// data buffer to store data before syncing and to read after syncing
+	private var dataBuffer = new HashBuffer
+
+	// encoder to send strings over the network
+	private var stringEncoder = new StringEncoder
+
+	// gamecache buffers to sync data in rounds
+	private var gcIntBuffer     = new GamecacheBuffer(DATA_CACHE, mkey, BufferType.INTEGER)
+	private var gcRealBuffer    = new GamecacheBuffer(DATA_CACHE, mkey, BufferType.REAL)
+	private var gcBooleanBuffer = new GamecacheBuffer(DATA_CACHE, mkey, BufferType.BOOLEAN)
+	private var gcStringBuffer  = new GamecacheBuffer(STRING_CACHE, mkey, BufferType.INTEGER)
+
+	// metadata container
+	private var metadataStore = new MetadataStore
+
+	// the player that we are receiving data from
+	private player sender
+
+	// metadata about the transmission
+	private var meta = metadata(0, 0, 0, 0, 0)
```

Comment:
make some EMPTY_META constant for the null element

Reply:
Done